### ISSUE TITLE
Failure to embed same nested association attribute on the root (Exception) 0.9-stable

### ISSUE THREAD

**Issue by josercruz01** [rails-api/active_model_serializers#1054]
There seems to be a small bug when embedding nested associations in the root. AMS is failing to merge the ```Type``` field of both the child ```User``` instance and the parent ```User``` instance on the root level.

I need to serialize the following association, but the serializer fails to merge the values for the nested ```TypeSerializer``` by throwing an exception inside the ```embedded_in_root_associations```:

```ruby
class TypeSerializer < ActiveModel::Serializer
  attributes :name, :id
end

class UserSerializer < ActiveModel::Serializer
  attributes :name, :id

  has_one :type, serializer: TypeSerializer, embed: :ids, include: true
  has_one :parent, serializer: UserSerializer, embed: :ids, include: true
end

child_type = Type.create(name: "type-child")
parent_type = Type.create(name: "type-parent")

parent = User.create(name: "The Parent", type: parent_type)
child1 = User.create(name: "The child", type: child_type, parent: parent)

result = UserSerializer.new(child1).as_json
```

That results on the following error:
```bash
NoMethodError: undefined method `merge' for [{:name=>"type-child", :id=>1}]:Array
from /opt/rubies/2.2.2/lib/ruby/gems/2.2.0/bundler/gems/active_model_serializers-dc08f0ec80ce/lib/active_model/serializer.rb:215:in `block (2 levels) in embedded_in_root_associations'
```

### Test
To better explain the issue I created a test on [this repository](https://github.com/josercruz01/serializer-test) (click here for [specific test case method](https://github.com/josercruz01/serializer-test/blob/master/test/controllers/user_serializer_test.rb)). To run the test just execute: ```rake test```. The code will run the same test case I placed above, you can see the exception on the console.

### Issue
At first sight, the issue seems to be that [the merge of inner associations embedded in root](https://github.com/rails-api/active_model_serializers/blob/0-9-stable/lib/active_model/serializer.rb#L209) assumes that the the key collision will be of type ```Hash```, but in this case the type is ```Array```, and it blows up because it's calling ```merge``` on ```Array```.

Issue is solved in this case if I do the following, but not sure if there is any other edge case where this might affect the result:

```ruby
hash.merge!(association_serializer.embedded_in_root_associations) do |key, oldval, newval|
  if oldval.respond_to?(:to_ary)
    [oldval, newval].flatten.uniq
  else
    oldval.merge(newval) { |_, oldval, newval| [oldval, newval].flatten.uniq }
  end
end
``` 

Any thoughts on this problem?

**Comment by thec0keman** [comment]
We have a very similar situation on 0.9-stable and this fix seems to resolve.

**Comment by fractaloop** [comment]
I too experienced this issue, but found @Ferdev's solution solves the issue. Please consider reviewing this fix. Thank you.

**Comment by bf4** [comment]
Thanks @fractaloop @thec0keman @josercruz01 @ferdev  Anyone want to make a PR? I'd appreciate it.

**Comment by RustComet** [comment]
Ahhh didn't see the @fractaloop PR before opening mine. Will close it now!

**Comment by Hirtenknogger** [comment]
same problem here.

**Comment by bf4** [comment]
@Hirtenknogger @thec0keman @josercruz01 @ferdev can y'all checkout https://github.com/rails-api/active_model_serializers/pull/1523  CI is having issues and I'm happy to merge and release if it works for you

**Comment by Hirtenknogger** [comment]
@bf4 I am not familiar with it. 
Would be great if someone else could check out @thec0keman @josercruz01 @ferdev .

**Comment by sagarbommidi** [comment]
Any updates on this issue pls? Am also facing the same issue.

**Comment by fractaloop** [comment]
I'm sorry y'all, but I don't have the time to add tests and stuff for this. I don't use Rails professionally and this was just for a side project. Would someone be willing to take the reins on this?

**Comment by bf4** [comment]
@sagarbommidi https://github.com/rails-api/active_model_serializers/issues/1054#issuecomment-194200468

**Comment by remear** [comment]
@Hirtenknogger It looks like @fractaloop is unable to continue with his PR. https://github.com/rails-api/active_model_serializers/pull/1523#issuecomment-198025303

Would you like to take the lead on a PR for this?

**Comment by Hirtenknogger** [comment]
@remear It is a problem with the configuration of appveyor. There is no appveyor.yml in the branch 0-9-stable. perhaps we can merge appveyor branch to fix it?!

**Comment by Hirtenknogger** [comment]
@bf4 @remear @sagarbommidi done.

**Comment by sagarbommidi** [comment]
:+1: 

That's awesome, thanks @Hirtenknogger .

**Comment by groyoh** [comment]
Fixed by #1607
